fontir/src/orchestration.rs

```diff
@@ -34,83 +32,306 @@ bitflags! {
+    pub fn try_get(&self) -> Option<Arc<T>> {
+        self.acl.assert_read_access(&self.id);
+        self.value.read().as_ref().cloned()
+    }
+
+    /// Update the value of the item whether or not it has changed.
+    ///
+    /// The change will be logged and anything relying on change detection, such as
+    /// conditional execution of dependent tasks, will fire.
+    ///
+    /// [ContextItem::set] is preferable where possible.
+    ///
+    /// This exists largely because write types in fontations do not always implement PartialEq.
+    /// TODO: should they?
+    pub fn set_unconditionally(&self, value: T) {
+        self.acl.assert_write_access(&self.id);
+
+        if self.irio.emit_ir() {
+            let mut writer = self.irio.writer(&self.id);
+            value.write(&mut writer);
+        }
+
+        *self.value.write() = Some(Arc::from(value));
+    }
+}
+
+impl<I, T, Ir> ContextItem<I, T, Ir>
+where
+    I: Clone + Debug + PartialEq + Eq + Hash,
+    T: PartialEq,
+    Ir: IrIo<I>,
+    T: Persistable,
+{
+    /// Update the value if it has changed.
+    ///
+    /// Change logging and dependent task execution will only fire if the value changed.
+    pub fn set(&self, value: T) {
+        self.acl.assert_write_access(&self.id);
+
+        // nop?
+        if self
+            .value
+            .read()
+            .as_ref()
+            .map(|arc| **arc == value)
+            .unwrap_or(false)
+        {
+            eprintln!("set_if_changed {:?}, NO CHANGE!", self.id);
+            return;
+        }
+
+        eprintln!("set_if_changed {:?}, changed!", self.id);
+        self.set_unconditionally(value);
+    }
+}
+
+/// Clones are cheap and reference the same wrapped item courtesy of Arc
 ///
-/// <https://veykril.github.io/tlborm/decl-macros/minutiae/fragment-specifiers.html>
-#[macro_export]
-macro_rules! context_accessors {
-    ($getter_name:ident, $setter_name:ident, $has_name:ident, $lock_name:ident, $value_type:ty, $id:expr, $restore_fn:ident, $prepersist_fn:ident) => {
-        pub fn $getter_name(&self) -> Arc<$value_type> {
-            let id = $id;
-            self.acl.assert_read_access(&id.clone().into());
-            {
-                let rl = self.$lock_name.read();
-                if rl.is_some() {
-                    return rl.as_ref().unwrap().clone();
-                }
-            }
-            set_cached(&self.$lock_name, $restore_fn(&self.paths.target_file(&id)));
-            let rl = self.$lock_name.read();
-            rl.as_ref().expect(MISSING_DATA).clone()
+/// Courtesy of Arc this is Clone even if T isn't
+#[derive(Default)]
+pub struct ContextMap<I, T, Ir>
+where
+    I: Clone + Debug + PartialEq + Eq + Hash,
+    T: IdAware<I>,
+    Ir: IrIo<I>,
+{
+    acl: Arc<AccessControlList<I>>,
+    irio: Arc<Ir>,
+    value: Arc<RwLock<HashMap<I, Arc<T>>>>,
+}
+
+impl<I, T, Ir> ContextMap<I, T, Ir>
+where
+    I: Clone + Debug + PartialEq + Eq + Hash,
```

Comment:
At this point I'm starting to wonder if we should just define,

```rust
pub trait Identifier: Clone + Debug + Eq + Hash {}
```
?

Reply:
Probably; it's a pretty common idea. Per your other comment perhaps debug isn't strictly necessary although ... I think being able to debug your ids is going to end up ubiquitous.

Reply:
I like it, conveys the intent well.